ev3dev2/motor.py

```diff
@@ -2182,6 +2189,138 @@ def turn_left(self, speed, degrees, brake=True, block=True):
+                right_ticks = right_current - right_previous
+
+                # Have we moved?
+                if not left_ticks and not right_ticks:
+                    time.sleep(SLEEP_TIME)
+                    continue
+
+                # log.debug("%s: left_ticks %s (from %s to %s)" %
+                #     (self, left_ticks, left_previous, left_current))
+                # log.debug("%s: right_ticks %s (from %s to %s)" %
+                #     (self, right_ticks, right_previous, right_current))
+
+                # update _previous for next time
+                left_previous = left_current
+                right_previous = right_current
+
+                # rotations = distance_mm/self.wheel.circumference_mm
+                left_rotations = float(left_ticks / self.left_motor.count_per_rot)
+                right_rotations = float(right_ticks / self.right_motor.count_per_rot)
+
+                # convert longs to floats and ticks to mm
+                left_mm = float(left_rotations * self.wheel.circumference_mm)
+                right_mm = float(right_rotations * self.wheel.circumference_mm)
+
+                # calculate distance we have traveled since last sampling
+                mm = (left_mm + right_mm) / 2.0
+
+                # accumulate total rotation around our center
+                self.theta += (left_mm - right_mm) / self.wheel_distance_mm
+
+                # and clip the rotation to plus or minus 360 degrees
+                self.theta -= float(int(self.theta/TWO_PI) * TWO_PI)
+
+                # now calculate and accumulate our position in mm
+                self.x_pos_mm += mm * math.sin(self.theta)
+                self.y_pos_mm += mm * math.cos(self.theta)
+
+                self.odometry_coordinates_log()
+                time.sleep(SLEEP_TIME)
+
+            self.odometry_thread_id = None
+
+        self.odometry_thread_run = True
+        self.odometry_thread_id = _thread.start_new_thread(_odometry_monitor, ())
+
+    def odometry_stop(self):
+        """
+        Signal the odometry thread to exit and wait for it to exit
+        """
+
+        if self.odometry_thread_id:
+            self.odometry_thread_run = False
+
+            while self.odometry_thread_id:
+                pass
+
+    def turn_to_angle(self, speed, angle_target_degrees, brake=True, block=True):
+
+        if angle_target_degrees < 0:
+            angle_target_degrees += 360
+
+        angle_current_degrees = math.degrees(self.theta)
+
+        if angle_current_degrees < 0:
+            angle_current_degrees += 360
+
+        angle_delta = angle_current_degrees - angle_target_degrees
+
+        log.info("%s: turn_to_angle %s, current angle %s, delta %s" %
+            (self, angle_target_degrees, angle_current_degrees, angle_delta))
+        self.odometry_coordinates_log()
+
+        # Do we rotate left or right and by how many degrees?
+        if angle_current_degrees > angle_target_degrees:
```

Comment:
I think this needs to do a more intelligent conversion by finding the shortest difference around the circle... What happens if I'm facing 5 degrees and request 355? Right now it looks like it will turn 350 degrees clockwise rather than 10 counterclockwise.

Reply:
sorry for the confusion, this wasn't ready for review. I was posting the patch to discuss the angle thing to try to get my head wrapped around that part.